File: DotNet/Messaging/HydraService.cs
using System;
using Bollywell.Hydra.Messaging.Listeners;
using Bollywell.Hydra.Messaging.MessageFetchers;
using Bollywell.Hydra.Messaging.MessageIds;
using Bollywell.Hydra.Messaging.Storage;

namespace Bollywell.Hydra.Messaging
{
    public class HydraService : IHydraService
    {
        private readonly IProvider _provider;

        public ListenerOptions DefaultListenerOptions { get; set; }

        public HydraService(IProvider provider, ListenerOptions defaultListenerOptions = null)
        {
            _provider = provider;
            DefaultListenerOptions = defaultListenerOptions;
        }

        #region Implementation of IHydraService

        public IListener<TMessage> GetListener<TMessage>(IMessageFetcher<TMessage> messageFetcher, IMessageId startId = null, ListenerOptions listenerOptions = null) where TMessage : TransportMessage
        {
            return new Listener<TMessage>(_provider, messageFetcher, startId, listenerOptions ?? DefaultListenerOptions);
        }

        public IMessageId Send<TMessage>(TMessage message) where TMessage : TransportMessage
        {
            bool failed = false;
            while (!failed && !_provider.IsOffline) {
                var store = _provider.GetStore(true);
                if (store == null) {
                    failed = true;
                } else {
                    try {
                        return store.SaveDoc(message.ToJson());
                    }
                    catch (Exception) {
                        _provider.ServerError(store.Name);
                    }
                }
            }
            throw new Exception("HydraService.Send: Error sending message - all servers offline.");
        }
        
        #endregion

    }
}

File: Java/hydra/src/main/java/uk/co/shastra/hydra/messaging/HydraService.java
package uk.co.shastra.hydra.messaging;

import java.io.InputStream;

import uk.co.shastra.hydra.messaging.attachments.Attachment;
import uk.co.shastra.hydra.messaging.listeners.Listener;
import uk.co.shastra.hydra.messaging.listeners.ListenerOptions;
import uk.co.shastra.hydra.messaging.messagefetchers.MessageFetcher;
import uk.co.shastra.hydra.messaging.messageids.MessageId;

public interface HydraService {
	/**
	 * Start listening for messages using the supplied MessageFetcher. 
	 * 
	 * @param messageFetcher The MessageFetcher with which to poll for messages.
	 * @return A Listener that makes incoming messages available.
	 */
	<TMessage extends TransportMessage> Listener<TMessage> getListener(MessageFetcher<TMessage> messageFetcher);
	/**
	 * Listen for messages since startId using the supplied MessageFetcher. 
	 * 
	 * @param messageFetcher The MessageFetcher with which to poll for messages.
	 * @param startId Fetch all messages on or after this MessageId
	 * @return A Listener that makes incoming messages available.
	 */
	<TMessage extends TransportMessage> Listener<TMessage> getListener(MessageFetcher<TMessage> messageFetcher, MessageId startId);
	/**
	 * Listen for messages since startId using the supplied MessageFetcher, and with the supplied ListenerOptions
	 * 
	 * @param messageFetcher The MessageFetcher with which to poll for messages.
	 * @param startId Fetch all messages on or after this MessageId
	 * @param listenerOptions Options to configure the Listener behaviour 
	 * @return A Listener that makes incoming messages available.
	 */
	<TMessage extends TransportMessage> Listener<TMessage> getListener(MessageFetcher<TMessage> messageFetcher, MessageId startId, ListenerOptions listenerOptions);

	/**
	 * Send a message.
	 * 
	 * @param message The message to send.
	 * @return The MessageId of the sent message.
	 * @throws Exception If there is an error sending the message.
	 */
	<TMessage extends TransportMessage> MessageId send(TMessage message) throws Exception;
	/**
	 * Send a message with attachments.
	 * 
	 * @param message The message to send.
	 * @param attachments Attachments to send with the message.
	 * @return The MessageId of the sent message.
	 * @throws Exception If there is an error sending the message.
	 */
	<TMessage extends TransportMessage> MessageId send(TMessage message, Iterable<Attachment> attachments) throws Exception;
	
	/**
	 * Get a message attachment.
	 * 
	 * @param attachment The Attachment to fetch.
	 * @return The attachment, as an InputStream.
	 * @throws Exception 
	 */
	InputStream getAttachment(Attachment attachment) throws Exception;
	
    /**
     * @return The name of the Hydra server currently in use.
     */
    String getServerName();
}
